include/seastar/http/client.hh

```diff
@@ -34,9 +34,23 @@ namespace tls { class certificate_credentials; }
 
 namespace http {
 
+namespace experimental { class client; }
 struct request;
 struct reply;
 
+namespace internal {
+
+class client_ref {
+    http::experimental::client* _c;
+public:
+    client_ref(http::experimental::client* c) noexcept;
+    ~client_ref();
+    client_ref(client_ref&& o) noexcept : _c(std::exchange(o._c, nullptr)) {}
+    client_ref(const client_ref&) = delete;
```

Comment:
I see you invent here a new counter mechanism, a condition variable to wait on it, and a "client_ref" to RAII-hold it. But Seastar already has a feature that (as far as I can tell) does all of this - semaphore, and its `wait()` and `get_units()`. Am I missing something that makes the new mechanism different somehow?

Reply:
Well, it's not a semaphore, the condition variable here doesn't wait for the counter to become non-zero (or go strictly less than some maximum value). Wake-up can happen when one of sockets (that hold a reference on client) gets into the idling list (a.k.a. pool).

Reply:
Another "pros" for the cond variable + counter is that maximum number of connections a client may have can be dynamically updated (I have this draft branch), which is probably good. Contrary to this, semaphore's maximum `_count` value is defined constructor-time and cannot (well, presumably shouldn't, `basic_semaphore::signal(int nr)` doesn't check it) change on the fly.

Reply:
@nyh , the updated version has the on-the-fly update call. Please, take another look at it, it should address your concern about using counter+condvar+raii_helper instead of a semaphore